Fully support event updates for cached message lists.

amanagr opened the pull request:
This will allow us to confidently maintain cached data structures which will show correct information.

amanagr committed:
message_events: Update cached msg list data for msg property update.

We extend the existing logic that updates message events for
rendered message lists to apply for cached message list data
objects as well.

Manually tested for common message property narrows to test
update of the message property.

amanagr commented:
Not ready for review.

timabbott commented on web/src/message_events.js:
```diff
@@ -47,6 +47,33 @@ function filter_has_term_type(filter, term_type) {
+    // We loop over rendered message lists and cached message data separately since
+    // they are separately maintained and can have different items.
+    for (const msg_list of message_lists.all_rendered_message_lists()) {
+        if (exclude_current_msg_list && msg_list === message_lists.current) {
```
Do we have a use case where we wouldn't want to exclude the current list? It seems like it should basically always be unsafe to discard that one.

amanagr replied:
No we don't have a use case for that, Thanks!

timabbott commented on web/src/message_events.js:
```diff
@@ -102,6 +102,152 @@ export function update_current_view_for_topic_visibility() {
+                msg_list_data.add_messages(messages_to_add);
+                msg_list?.rerender();
+            },
+        });
```
We might want to pass an option that lets us consider discarding these from the cache rather than fetching from the server if it's not the current view? 

What seems dangerous about this implementation is that if your last 100 views were all search views, and some messages get moved, this could trigger 100 requests to the server to fetch whether those messages match all of those searches... 

Though do we even currently keep cached `!can_apply_locally` data lists? Maybe the right answer is to ensure we don't do that, and have a strategically placed `assert`.

amanagr replied:
Okay, discarded the cached data if it is not part of a rendered message list.

I think we can store `!can_apply_locally` datasets and use them to render narrows. See the commit
`message_view: Attempt rendering cached data without contacting server.`.

timabbott replied:
OK -- that's fine as long as we have a way to discard those when we cannot ensure they remain accurate.

timabbott commented on web/src/fetch_status.ts:
```diff
@@ -141,4 +141,10 @@ export class FetchStatus {
+    copy_status(fetch_status: FetchStatus): void {
+        this._found_newest = fetch_status.has_found_newest();
+        this._found_oldest = fetch_status.has_found_oldest();
+        this._expected_max_message_id = fetch_status._expected_max_message_id;
```
Should `_history_limited` be included? I think you need a comment on the `loading` ones as to why they aren't set here.

amanagr replied:
Yes, somehow I missed that.

timabbott commented:
I didn't try to read the WIP commit, but reviewed the earlier stuff with a view towards supporting merging prep commits.

amanagr committed:
message_view: Copy fetch status of superset data to msg data.

We can rely on the fetch status of the superset data since it
is now live updated for message events.

amanagr committed:
message_view: Attempt rendering cached data without contacting server.

For narrows for which we cannot filter the messages locally,
we attempt to select a message without contacting the server.

amanagr committed:
message_view: Avoid initial fetch if we have enough context to render.

We don't contact the server at all when rendering a narrow if the
locally cached data has enough number of messages.

amanagr committed:
message_events: Update cached msg list data when moving messages.

amanagr commented on web/src/message_fetch.ts:
```diff
@@ -103,6 +103,40 @@ export function load_messages_around_anchor(
+    if (
+        !message_lists.current.is_combined_feed_view &&
+        message_lists.current.visibly_empty()
+    ) {
```
Done.

amanagr committed:
message_events: Fix live update for search narrow on content update.

For messages whose content was edited, we ask the server if the
message matches the current search narrow.

timabbott commented on web/src/echo.ts:
```diff
@@ -560,6 +561,16 @@ export function process_from_server(messages: ServerMessage[]): ServerMessage[]
+            if (!msg_list_data.filter.can_apply_locally()) {
+                // Ideally we would ask server to if messages matches filter
+                // but it is not worth doing so for every new message.
+                message_list_data_cache.remove(msg_list_data.filter);
```
This is probably a case where we could refine the logic to be more precise -- if other operators on the `filter` mean the message can't match, we could not discard. And we could potentially extend the `MessageListData` structure with a list of messages to check before rendering it.

I'll open a follow-up issue, since most searches will just have the keyword operator anyway.

amanagr replied:
Yeah, makes sense.

timabbott commented on web/src/server_events_dispatch.js:
```diff
@@ -1022,6 +1023,14 @@ export function dispatch_normal_event(event) {
+                user_topics.get_topic_visibility_policy(event.stream_id, event.topic_name) ===
+                    user_topics.all_visibility_policies.FOLLOWED
+            ) {
+                message_events.discard_cached_lists_with_term_type("is-followed");
```
Does this second condition work? I think you need to do this before `user_topics_ui.handle_topic_updates`, or at least before `user_topics.set_user_topic(user_topic_event);` at the start of that function (which could reasonably be pulled up to sit here if helpful)

timabbott replied:
This seems correct; please hand-test though:

``` diff
diff --git a/web/src/server_events_dispatch.js b/web/src/server_events_dispatch.js
index f94f50d0fa..7f99f99a1d 100644
--- a/web/src/server_events_dispatch.js
+++ b/web/src/server_events_dispatch.js
@@ -1018,7 +1018,11 @@ export function dispatch_normal_event(event) {
             }
             break;
 
-        case "user_topic":
+        case "user_topic": {
+            const previous_topic_visibility = user_topics.get_topic_visibility_policy(
+                event.stream_id,
+                event.topic_name,
+            );
             user_topics_ui.handle_topic_updates(
                 event,
                 message_events.update_current_view_for_topic_visibility(),
@@ -1026,11 +1030,11 @@ export function dispatch_normal_event(event) {
             // Discard cached message lists if `event` topic was / is followed.
             if (
                 event.visibility_policy === user_topics.all_visibility_policies.FOLLOWED ||
-                user_topics.get_topic_visibility_policy(event.stream_id, event.topic_name) ===
-                    user_topics.all_visibility_policies.FOLLOWED
+                previous_topic_visibility === user_topics.all_visibility_policies.FOLLOWED
             ) {
                 message_events.discard_cached_lists_with_term_type("is-followed");
             }
             break;
+        }
     }
 }
```

amanagr replied:
oh yeah!

timabbott commented on web/tests/narrow_activate.test.js:
```diff
@@ -1,292 +0,0 @@
-
-const {mock_esm, set_global, zrequire} = require("./lib/namespace");
-const {run_test, noop} = require("./lib/test");
-const $ = require("./lib/zjquery");
```
Yeah in theory a good node test for this code path would be good, but the coverage of actual logic is quite poor (better replaced with unit tests for individual functions like `maybe_add_local_messages` if we wanted to invest time in it) , and it's so darn much mocking.

timabbott commented on web/src/message_events.js:
```diff
@@ -660,6 +660,8 @@ export function update_messages(events) {
                 //       when we move to new narrow and what data is locally available.
                 if (changed_narrow) {
+                    // Remove outdated cached data to avoid repopulating from it.
+                    message_list_data_cache.remove(new_filter);
```
Can you expand the comment here with more reasoning on why this is correct or under what circumstances we could change the logic? In theory, we're changing the narrow because the messages moved, but if we were able to live-update the `MessageListData` for a cached destination narrow for the move, it should be fine. Is the issue that we've not attempted to do such a live-update of the `MessageListData` objects yet in the sequence in this file?

amanagr replied:
Yeah, the main issue is we have't updated the cached data yet. Added a comment for that.

timabbott commented on web/src/message_view.ts:
```diff
@@ -974,6 +974,30 @@ export function maybe_add_local_messages(opts: {
+            msg_data.fetch_status.has_found_oldest()
+        ) {
+            id_info.local_select_id = msg_data.first_unread_message_id();
+        }
```
Hmm, can you explain why this logic is being added to the `cannot_compute` block? I'm not sure what assumptions are being made here that justify being able to navigate the user correctly. (I could imagine maybe that we should instead be not returning `cannot_compute` if we have the view cached and can somehow use that to pin the right message ID to select).

amanagr replied:
So, here if we have the `MessageListData` with the same filter as we are trying to render, we can try to render the
message list locally.

We have a `target_id` we want to select - Likely user searched for something -> clicked on a messsage and pressed back key. Then, we can select the message.

We don't have a `target_id` to select but have the complete message history for the filter. In that case we can just render and select the first unread.

timabbott replied:
OK. I may want to add some comments here.

amanagr replied:
Done

timabbott commented on web/src/message_view.ts:
```diff
@@ -692,7 +692,29 @@ export function show(raw_terms: NarrowTerm[], show_opts: ShowMessageViewOpts): v
+                        has_found_newest,
+                        true,
+                        true,
+                    );
```
This logic seems like it needs to consider that if both `has_found_oldest` and `has_found_newest`, we should always have enough context regardless of the number of messages, right? (Like a topic with 3 messages total will always fetch as written here, I think?)

amanagr replied:
Yeah, adjusted the logic to exclude fetch for that case.

timabbott commented on web/src/message_events.js:
```diff
@@ -884,8 +884,31 @@ export function update_messages(events) {
+                    messages_to_rerender,
+                    msg_list,
+                    message_util.add_new_messages,
+                );
```
Does this need to consider the possible of the edit removing the message from the view?

amanagr replied:
Yup! Added a commit for that which I tested locally by editing a message to remove it from the search.

timabbott commented:
I reviewed most of this and opened #32259 to merge some of the obviously safe preparatory refactors; I'm not entirely sure how much more of a prefix of this doesn't introduce temporary regressions, so leaving the rest as-is for now. It's probably reasonable to test-deploy on chat.zulip.org and start seeing if we can find bugs after the next revision.

amanagr commented:
(Accidentally deleted the comment)

timabbott commented on web/src/message_view.ts:
```diff
@@ -692,7 +692,30 @@ export let show = (raw_terms: NarrowTerm[], show_opts: ShowMessageViewOpts): voi
+                        true,
+                        true,
+                    );
+                }
```
I think the fallthrough case might be bugged here; the `else` clause logic doesn't run if the above if statement failed to apply.

amanagr replied:
We actually don't want to contact server if we have complete history or enouch context because in that case, we will automatically fetch more messages if required due to this logic in `scroll_finished` block:

```js

    if (message_lists.current.view.should_fetch_older_messages()) {
        // Subtle note: While we've only checked that we're at the
        // very top of the render window (i.e. there may be some more
        // cached messages to render), it's a good idea to fetch
        // additional message history even if we're not actually at
        // the edge of what we already have from the server.
        message_fetch.maybe_load_older_messages({
            msg_list: message_lists.current,
            msg_list_data: message_lists.current.data,
        });
    }

    if (message_lists.current.view.should_fetch_newer_messages()) {
        // See the similar message_viewport.at_rendered_top block.
        message_fetch.maybe_load_newer_messages({
            msg_list: message_lists.current,
        });
    }
```

amanagr replied:
Added a comment here.

timabbott replied:
OK, thanks, the comment is very helpful. I started a #design thread about if we want to do some double-checking fetches.

timabbott commented:
Deployed on chat.zulip.org for testing.

timabbott commented:
We at least haven't noticed any anomalies yet. This probably will need some bake time though -- and probably we should look at adding the logging of discrepancies.

amanagr commented:
> We at least haven't noticed any anomalies yet. This probably will need some bake time though -- and probably we should look at adding the logging of discrepancies.

Yeah, I am working on it.

amanagr commented:
> We at least haven't noticed any anomalies yet. This probably will need some bake time though -- and probably we should look at adding the logging of discrepancies.

Yeah, I am working on it.

amanagr committed:
message_view: Verify msg list data state when restoring cached data.

When rendering message list directly from cached data, we sometimes
ask server to verify the current data.

If the data doesn't match, we include events at the time cached
data was restored to help with debugging or to determine if there
was race in updating the data.

amanagr commented:
> We at least haven't noticed any anomalies yet. This probably will need some bake time though -- and probably we should look at adding the logging of discrepancies.

Added a commit for that as per discussion [here](https://chat.zulip.org/#narrow/channel/101-design/topic/Improved.20use.20of.20cached.20message.20history).

timabbott commented:
OK updating the deployment.

amanagr committed:
message_fetch: Account for recently sent messages in data verification.

amanagr commented:
Moved the constant to `message_fetch.const`.

amanagr committed:
message_fetch: Don't include full messages data in sentry logs.

Sentry has a limit for how much data it can store as extra info,
we cannot pass full message object here.

amanagr committed:
message_view: Fix incomplete msg history for new users.

Since new users likely don't have UserMessage for all the messages,
we need to check if the superset_data includes complete history
independent of UserMessage or not when populating narrows
which don't depend on UserMessage presence.

amanagr commented on web/src/message_view.ts:
```diff
@@ -948,9 +949,39 @@ function load_local_messages(msg_data: MessageListData, superset_data: MessageLi
+    // `zerver.lib.narrow.ok_to_include_history` is not `True` when
+    // populating narrow for which it is `True` since older
+    // messages might be filtered due to them not having `UserMessage`
+    // entries. See `zerver.lib.narrow.fetch_messages`.
```
Also, verified that we get `BlueslipError: Mismatching cached and server data.` without this commit when narrowing to a topic with more message than in the combined feed.

amanagr commented:
The CI failure is unrelated to the changes I made here:

`sqlalchemy.exc.SADeprecationWarning: Invoking and_() without arguments is deprecated, and will be disallowed in a future release.   For an empty and_() construct, use and_(True, *args).`

timabbott commented on web/src/message_view.ts:
```diff
@@ -948,9 +949,39 @@ function load_local_messages(msg_data: MessageListData, superset_data: MessageLi
+        msg_data.fetch_status.copy_status(superset_data.fetch_status);
+    }
+
+    return true;
```
OK, this works. Somehow I'd thought we needed to do something more expensive to make this possible.

timabbott replied:
Oh, but I guess this is still not entirely correct, in that we don't handle the case that the current user was not continuously subscribed to the channel -- so the `all_messages_data` cache might be incomplete for that reason.

amanagr replied:
Oh yeah! I guess we will have to implement `never_unsubscribed` flag for channels as you mentioned.

timabbott commented:
Delete message puppeteer tests I've not seen fail elsewhere, so maybe related to this PR?

amanagr commented on web/src/message_view.ts:
```diff
@@ -913,8 +949,37 @@ function load_local_messages(msg_data: MessageListData, superset_data: MessageLi
+        ok_to_copy_status = false;
+    }
+
+    if (!ok_to_copy_status && msg_data.filter.has_operator("channel")) {
+        // We can still rely on fetch status if the superset_data
+        // has the first message of the channel.
+        const channel_id_str = msg_data.filter.operands("channel")[0];
+        assert(channel_id_str !== undefined);
+        const channel_id = Number.parseInt(channel_id_str, 10);
+        const oldest_possible_msg_id = sub_store.get(channel_id)?.first_message_id;
+        if (oldest_possible_msg_id && superset_data.get(oldest_possible_msg_id)) {
+            ok_to_copy_status = true;
+        }
+    }
```
@timabbott I think we can merge this PR by removing this optimization. 

Our blocker was this:

> Oh, but I guess this is still not entirely correct, in that we don't handle the case that the current user was not continuously subscribed to the channel -- so the all_messages_data cache might be incomplete for that reason.

The `if` condition above this will prevent populating the `msg_data` from `all_messages_data` in that case.

amanagr replied:
Past czo discussion - https://chat.zulip.org/#narrow/channel/3-backend/topic/Handling.20users.20being.20unsubscribed.20to.20a.20channel.20for.20a.20while

amanagr replied:
Let me test this and see if it fixes the bug.

amanagr replied:
Yeah, that fixes the issue.

amanagr replied:
Sorry, I didn't test completely, it would still won't show messages which were send while user was unsubscribed if a message was sent after user subscribed.

amanagr commented:
This needs a topic model to be correct. So, we can revisit after that.
